flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/restore/ForStIncrementalRestoreOperation.java

```diff
@@ -0,0 +1,393 @@
+                chooseMainHandleAndCollectOthers(otherHandles);
+
+        StateHandleTransferSpec mainSpec =
+                new StateHandleTransferSpec(
+                        mainHandle, new Path(forstBasePath, ForStResourceContainer.DB_DIR_STRING));
+
+        List<StateHandleTransferSpec> otherSpecs =
+                otherHandles.stream()
+                        .map(
+                                handle ->
+                                        new StateHandleTransferSpec(
+                                                handle,
+                                                new Path(
+                                                        forstBasePath,
+                                                        UUID.randomUUID().toString())))
+                        .collect(Collectors.toList());
+
+        try {
+            runAndReportDuration(
+                    () -> transferAllStateHandles(mainSpec, otherSpecs),
+                    // TODO: use new metric name, such as "TransferStateDurationMs"
+                    DOWNLOAD_STATE_DURATION);
+
+            runAndReportDuration(
+                    () -> restoreFromTransferredHandles(mainSpec, otherSpecs),
+                    RESTORE_STATE_DURATION);
+
+            return new ForStRestoreResult(
+                    this.forstHandle.getDb(),
+                    this.forstHandle.getDefaultColumnFamilyHandle(),
+                    this.forstHandle.getNativeMetricMonitor(),
+                    lastCompletedCheckpointId,
+                    backendUID,
+                    restoredSstFiles);
+        } finally {
+            // Delete the transfer destination quietly.
+            otherSpecs.stream()
+                    .map(StateHandleTransferSpec::getTransferDestination)
+                    .forEach(
+                            dir -> {
+                                try {
+                                    FileSystem fs = dir.getFileSystem();
+                                    fs.delete(dir, true);
+                                } catch (IOException ignored) {
+
+                                }
+                            });
+        }
+    }
+
+    private IncrementalRemoteKeyedStateHandle chooseMainHandleAndCollectOthers(
+            final List<IncrementalRemoteKeyedStateHandle> otherHandlesCollector) {
+        IncrementalRemoteKeyedStateHandle mainHandle = null;
+
+        // Just choose the handle with the biggest size
+        for (IncrementalRemoteKeyedStateHandle handle : restoreStateHandles) {
+            if (mainHandle == null) {
+                mainHandle = handle;
+            } else {
+                if (handle.getStateSize() > mainHandle.getStateSize()) {
+                    otherHandlesCollector.add(mainHandle);
+                    mainHandle = handle;
+                } else {
+                    otherHandlesCollector.add(handle);
+                }
+            }
+        }
```

Comment:
Why not do something like `RocksDBIncrementalCheckpointUtils.findTheBestStateHandleForInitial` does?

Reply:
Thanks for this good suggestions, left a todo, this pr focus on restore without parallelism change.